waspc/e2e-test/ShellCommands.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module ShellCommands
+  ( ShellCommand,
+    ShellCommandContext (..),
+    ShellCommandBuilder (..),
+    OutputDir (..),
+    runShellCommandBuilder,
+    combineShellCommands,
+    cdIntoCurrentProject,
+    appendToWaspFile,
+    reformatPackageJson,
+    setDbToPSQL,
+    waspCliNew,
+    waspCliCompile,
+    waspCliMigrate,
+    waspCliBuild,
+  )
+where
+
+import Control.Monad.Reader (MonadReader (ask), Reader, runReader)
+import Data.List (intercalate)
+import System.FilePath (joinPath, (</>))
+
+-- NOTE: Should we consider separating shell parts, Wasp CLI parts, and test helper parts in the future?
+--       This would likely reqiure adoption of some library, and creating new layers of abstraction.
+--       Deemed not worth it right now by all.
+
+-- NOTE: Using `wasp-cli` herein so we can assume using latest `stack install` in CI and locally.
+
+-- TODO: In future, find a good way to test `wasp-cli start`.
+
+type ShellCommand = String
+
+-- Each shell command gets access to the current project name, and maybe other things in future.
+data ShellCommandContext = ShellCommandContext
+  { _ctxtCurrentProjectName :: String
+  }
+  deriving (Show)
+
+-- Used to construct shell commands, while still giving access to the context (if needed).
+newtype ShellCommandBuilder a = ShellCommandBuilder {_runShellCommandBuilder :: Reader ShellCommandContext a}
+  deriving (Functor, Applicative, Monad, MonadReader ShellCommandContext)
+
+runShellCommandBuilder :: ShellCommandBuilder a -> ShellCommandContext -> a
+runShellCommandBuilder shellCommandBuilder context =
+  runReader (_runShellCommandBuilder shellCommandBuilder) context
+
+combineShellCommands :: [ShellCommand] -> ShellCommand
+combineShellCommands = intercalate " && "
+
+cdIntoCurrentProject :: ShellCommandBuilder ShellCommand
+cdIntoCurrentProject = do
+  context <- ask
+  return $ "cd " ++ _ctxtCurrentProjectName context
+
+appendToWaspFile :: String -> ShellCommandBuilder ShellCommand
+appendToWaspFile content =
+  -- NOTE: Using `show` to preserve newlines in string.
+  return $ "echo " ++ show content ++ " >> main.wasp"
+
+data OutputDir = DevOutputDir | BuildOutputDir
+
+-- Uses python2's json.tool to sort keys as `npm install` overwrites
```

Comment:
This is a pretty flaky dependency - python 2 is not being maintained anymore, may not be on all platforms, and we're depending on a whole different programming language. It works for now, but this should really be replaced with something more reliable in the future - I wonder whether Aeson has an ordering concept for instance.

Reply:
It appears this https://hackage.haskell.org/package/aeson-pretty-0.8.9/docs/Data-Aeson-Encode-Pretty.html
has a way to sort keys (confCompare) that can be defined.

Reply:
There are sadly a few fragile things in here we can improve, I'll add a more explicit TODO around that part 👍🏻